[nested tensor]add split and layer_norm_backward operations

YuqingJ opened the pull request:
Summary:
Add split and layer_norm_backward.

Note: It is non trivial to support split_with_sizes backward so adding the split operation to support the use case in the model.

Test Plan: unit tests

Differential Revision: D51052966

YuqingJ commented:
@pytorchbot label "topic: not user facing"

soulitzer commented on torch/nested/_internal/ops.py:
```diff
@@ -472,6 +472,31 @@ def prod_dim_int(func, *args, **kwargs):
+    values = inp._values
+
+    # hack to split on the last dim
+    dim = new_kwargs["dim"]
```
You can use `new_kwargs["dim"] = _wrap_jagged_dim(inp.ndim(), new_kwargs["dim"], "split")` to normalize the dim.

soulitzer commented on torch/nested/_internal/ops.py:
```diff
@@ -472,6 +472,31 @@ def prod_dim_int(func, *args, **kwargs):
+    split_size = new_kwargs["split_size"]
+    return tuple(
+        NestedTensor(values=x, **extract_kwargs(inp))
+        for x in torch.split(values, split_size, -1)
```
```suggestion
        for x in func(inp._values, **new_kwargs)
```
nit: no need to do `split_size = new_kwargs["split_size"]` etc, you can just use `new_kwargs` directly

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -3270,6 +3293,25 @@ def test_unbind(self, device):
+    def test_layer_norm(self, device):
+        a = torch.randn(2, 3, requires_grad=True, dtype=torch.float64, device=device)
+        b = torch.randn(3, 3, requires_grad=True, dtype=torch.float64, device=device)
+        c = torch.randn(4, 3, requires_grad=True, dtype=torch.float64, device=device)
```
FYI, we recently added this new helper so you can do something like `ts = self._get_list_for_jagged_tensor(((2, 3, 4), 3), device=device, requires_grad=True)`

soulitzer commented on torch/nested/_internal/ops.py:
```diff
@@ -730,6 +755,20 @@ def native_layer_norm_default(func, *args, **kwargs):
+    grad_out = new_kwargs.pop("grad_out")
+    inp = new_kwargs.pop("input")
+    d_input, d_gamma, d_beta = func(grad_out._values, inp._values, **new_kwargs)
+    return (NestedTensor(d_input, **extract_kwargs(inp)), d_gamma, d_beta)
```
We should check if `d_input` is None here in the case where input does not require grad.

soulitzer reviewed (commented):
Thanks for the addition, mostly minor comments.
One thing to note for backward functions is that they take a `output_mask` which indicates which inputs actually require grad. If a particular input does not require grad, the backward function should return None, and we should handle that specially.

YuqingJ committed:
[nested tensor]add split and layer_norm_backward operations (#113108)

Summary:

Add split and layer_norm_backward.

Note: It is non trivial to support split_with_sizes backward so adding the split operation to support the use case in the model.

Test Plan: unit tests

Differential Revision: D51052966

soulitzer reviewed (approved):
Thanks!

YuqingJ commented:
@pytorchbot merge
